File: singly_linked_list.java
class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class linked_list {
    Node head;

    linked_list() {
        this.head = null;
    }

    void insert_at_beginning(int data) {
        Node new_node = new Node(data);
        if (this.head == null) {
            this.head = new_node;
        } else {
            new_node.next = this.head;
            this.head = new_node;
        }
    }

    void insert_in_between(int data, int key) {
        Node new_node = new Node(data);
        if (this.head == null) {
            this.head = new_node;
            return;
        }
        Node temp = this.head;
        while (temp != null) {
            if (temp.data == key) {
                new_node.next = temp.next;
                temp.next = new_node;
                return;
            }
            temp = temp.next;
        }
    }

    void insert_at_end(int data) {
        Node new_node = new Node(data);
        if (this.head == null) {
            this.head = new_node;
            return;
        }
        Node temp = this.head;
        while (temp.next != null) {
            temp = temp.next;
        }
        temp.next = new_node;
    }

    void delete_node(int key) {
        Node prev = null;
        Node temp = this.head;
        if (temp != null && temp.data == key) {
            this.head = temp.next;
            return;
        }
        while (temp != null && temp.data != key) {
            prev = temp;
            temp = temp.next;
        }
        if (temp == null) {
            return; // Key not found
        }
        prev.next = temp.next;
    }

    void print_linked_list() {
        Node temp = this.head;
        while (temp != null) {
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.println();
    }
}

class Main {
    public static void main(String[] args) {
        linked_list list = new linked_list();
        list.insert_at_beginning(20);
        list.insert_at_beginning(25);
        list.insert_in_between(56, 20);
        list.insert_at_end(45);

        list.print_linked_list();
    }
}

File: singly_linked_list.c
#include <stdio.h>
#include <stdlib.h>

struct Node{
    int data;
    struct Node* next;
};

struct Node* create_node(int data){
    struct Node* node=(struct Node*)malloc(sizeof(struct Node));
    node->data=data;
    node->next=NULL;
    return node;
}

struct Node** create_linked_list(){
    struct Node** head = malloc(sizeof(struct Node*));
    *head = NULL;
    return head;
}

int insert_at_beginng(struct Node** head, int data){
    struct Node* new_node=create_node(data);
    if (*head==NULL){
        *head=new_node;
    }else{
        new_node->next=*head;
        *head=new_node;
    }
    return 0;
}
int insert_in_between(struct Node** head, int data,int key){
    struct Node* new_node=create_node(data);
    
    if (*head==NULL){
        *head=new_node;
        return 0;
    }
    struct Node* temp=*head;
    while(temp!=NULL){
        if (temp->data==key){
            new_node->next=temp->next;
            temp->next=new_node;
            return 0;
        }
        temp=temp->next;
    }
        return -1;
}
int insert_at_end(struct Node** head,int data){
    struct Node* new_node = create_node(data);
    if (*head==NULL){
        *head=new_node;
    }else{
        struct Node* temp=*head;
        while(temp->next!=NULL){
            temp=temp->next;
        }
        temp->next=new_node;
    }
    return 0;
}


int delete_node(struct Node** head, int key){
    struct Node* prev=NULL;
    struct Node* temp=*head;
    if (temp!=NULL && temp->data==key){
        *head=temp->next;
        free(temp);
        return 0;
    }
    while(temp->next!=NULL){
        if (temp->data==key){
            break;
        }
        prev=temp;
        temp=temp->next;
    }
    if (temp->next==NULL){
        if (temp->data==key){ // to deleate last element if key was at last
            prev->next=NULL;
            free(temp);
            return 0;
        }
        return -1;
    }
    prev->next=temp->next;
    free(temp);
    return 0;
}
int print_linked_list(struct Node** head){
    struct Node* temp;
    temp = *head;
    while(temp != NULL){
        printf("%d ", temp->data);
        temp = temp->next;
    }
    printf("\n");
    return 0;
}
int main(){
    struct Node** head=create_linked_list();
    insert_at_beginng(head,20);
    insert_at_beginng(head,25);
    insert_in_between(head,56,20);
    insert_at_end(head,45);
    
    
    print_linked_list(head);
    struct Node* temp=*head;
    int key =20;
    while (temp!=NULL){
        if (temp->data==20){
            break;
        }
        temp=temp->next;
        
    }
    printf("\n");
}


